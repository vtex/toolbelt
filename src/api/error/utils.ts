import { ErrorReport } from './ErrorReport'
import { ErrorKinds } from './ErrorKinds'

/** Create and register on telemetry an ErrorReport with errorKind = ErrorKinds.FLOW_ISSUE_ERROR */
export const createFlowIssueError = (msg: string): ErrorReport => {
  return ErrorReport.createAndMaybeRegisterOnTelemetry({
    kind: ErrorKinds.FLOW_ISSUE_ERROR,
    originalError: new Error(msg),
  })
}
