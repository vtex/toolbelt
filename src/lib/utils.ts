export function truncateStringsFromObject(element: any, maxStrSize: number) {
  if (element === null || element === undefined) {
    return element
  }
  if (typeof element === 'object') {
    Object.keys(element).forEach(key => {
      element[key] = truncateStringsFromObject(element[key], maxStrSize)
    })
    return element
  }
  if (typeof element === 'string' && element.length > maxStrSize) {
    return `${element.substr(0, maxStrSize)}[...TRUNCATED]`
  }
  return element
}
