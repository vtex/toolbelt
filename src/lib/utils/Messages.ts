import { ColorifyConstants } from '../constants/Colors'
import supportsHyperlinks from 'supports-hyperlinks'
import ansiEscapes from 'ansi-escapes'

export const formatHyperlink = (text: string, url: string): string =>
  supportsHyperlinks.stdout
    ? `${ColorifyConstants.URL_INTERACTIVE(ansiEscapes.link(text, url))}`
    : `${text} (${ColorifyConstants.URL_INTERACTIVE(url)})`
