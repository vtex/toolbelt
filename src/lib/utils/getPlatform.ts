import isDocker from 'is-docker'
import isWSL from 'is-wsl'

export function getPlatform() {
  if (isWSL) {
    return `${process.platform}:wsl`
  }

  if (isDocker()) {
    return `${process.platform}:container`
  }

  return process.platform
}
