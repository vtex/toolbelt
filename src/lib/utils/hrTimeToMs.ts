export function hrTimeToMs(hrtime: [number, number]) {
  return 1e3 * hrtime[0] + hrtime[1] / 1e6
}
