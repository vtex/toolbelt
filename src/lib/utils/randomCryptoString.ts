import { randomBytes } from 'crypto'

export function randomCryptoString(stringLength: number, alphabet: string) {
  if (alphabet.length > 256) {
    throw new Error("Argument 'alphabet' can't have more than 256 characters")
  }

  const bytes = randomBytes(stringLength)
  const result = new Array(stringLength)

  let cursor = 0
  for (let i = 0; i < stringLength; i++) {
    cursor += bytes[i]
    result[i] = alphabet[cursor % alphabet.length]
  }

  return result.join('')
}
