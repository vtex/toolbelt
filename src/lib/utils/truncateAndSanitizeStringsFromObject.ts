const tokenFields = ['authorization', 'authtoken', 'auth', 'token']

export function sanitizeJwtToken(token: string): string {
  const tokenParts = token.split('.')

  // Not JWT token
  if (tokenParts.length !== 3) {
    return token
  }

  return `${tokenParts[0]}.${tokenParts[1]}`
}

export function truncateAndSanitizeStringsFromObject(
  element: any,
  maxStrSize: number,
  depth = 3,
  seenObjects: any[] = [],
  currentKey: string = null
) {
  if (element == null) {
    return element
  }

  const objType = typeof element

  if (objType === 'object') {
    if (seenObjects.includes(element)) {
      return '[circular]'
    }

    if (Buffer.isBuffer(element)) {
      return { type: 'buffer', byteLength: Buffer.byteLength(element) }
    }

    if (depth === 0) {
      return Array.isArray(element) ? `[array]` : `[object]`
    }

    seenObjects.push(element)
    Object.keys(element).forEach(key => {
      // seenObjects.slice creates a copy of seenObjects at the current state
      element[key] = truncateAndSanitizeStringsFromObject(element[key], maxStrSize, depth - 1, seenObjects.slice(), key)
    })

    return element
  }

  if (objType === 'string') {
    const currentKeyNormalized = currentKey?.toLowerCase()
    if (tokenFields.includes(currentKeyNormalized)) {
      element = sanitizeJwtToken(element)
    }

    return element.length <= maxStrSize ? element : `${element.substr(0, maxStrSize)}[...TRUNCATED]`
  }

  if (objType === 'number' || objType === 'bigint' || objType === 'boolean' || objType === 'symbol') {
    return element
  }

  if (objType === 'function') {
    return `[function: ${element.name || 'anonymous'}]`
  }

  return `[${objType}]`
}
