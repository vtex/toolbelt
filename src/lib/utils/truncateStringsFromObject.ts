export function truncateStringsFromObject(element: any, maxStrSize: number, depth = 3, seenObjects: any[] = []) {
  if (element == null) {
    return element
  }

  const objType = typeof element

  if (objType === 'object') {
    if (seenObjects.includes(element)) {
      return '[circular]'
    }

    if (depth === 0) {
      return Array.isArray(element) ? `[array]` : `[object]`
    }

    seenObjects.push(element)
    Object.keys(element).forEach(key => {
      // seenObjects.slice creates a copy of seenObjects at the current state
      element[key] = truncateStringsFromObject(element[key], maxStrSize, depth - 1, seenObjects.slice())
    })

    return element
  }

  if (objType === 'string') {
    return element.length <= maxStrSize ? element : `${element.substr(0, maxStrSize)}[...TRUNCATED]`
  }

  if (objType === 'number' || objType === 'bigint' || objType === 'boolean' || objType === 'symbol') {
    return element
  }

  if (objType === 'function') {
    return `[function: ${element.name || 'anonymous'}]`
  }

  return `[${objType}]`
}
