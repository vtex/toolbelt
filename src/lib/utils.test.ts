import { truncateStringsFromObject } from './utils'

it.each([
  [5, ['123456', '12345', '1234']],
  [5, { a: '123456', b: '12345', c: ['123456', { a: '123456' }] }],
  [10, undefined],
  [10, null],
  [10, { a: undefined, b: null }],
  [
    5,
    {
      a: [{ aa: '123456' }, { ab: '1234' }, { ac: '12345' }],
      b: { ba: undefined, bb: null, bc: '123456', bd: { bca: '123456', bcb: '1234567' } },
    },
  ],
])('Works on testcase %# - String limit %d', (limit: number, obj: any) => {
  expect(truncateStringsFromObject(obj, limit)).toMatchSnapshot()
})
