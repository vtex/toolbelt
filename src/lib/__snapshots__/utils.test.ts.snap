// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Works on testcase 0 - String limit 5 1`] = `
Array [
  "12345[...TRUNCATED]",
  "12345",
  "1234",
]
`;

exports[`Works on testcase 1 - String limit 5 1`] = `
Object {
  "a": "12345[...TRUNCATED]",
  "b": "12345",
  "c": Array [
    "12345[...TRUNCATED]",
    Object {
      "a": "12345[...TRUNCATED]",
    },
  ],
}
`;

exports[`Works on testcase 2 - String limit 10 1`] = `undefined`;

exports[`Works on testcase 3 - String limit 10 1`] = `null`;

exports[`Works on testcase 4 - String limit 10 1`] = `
Object {
  "a": undefined,
  "b": null,
}
`;

exports[`Works on testcase 5 - String limit 5 1`] = `
Object {
  "a": Array [
    Object {
      "aa": "12345[...TRUNCATED]",
    },
    Object {
      "ab": "1234",
    },
    Object {
      "ac": "12345",
    },
  ],
  "b": Object {
    "ba": undefined,
    "bb": null,
    "bc": "12345[...TRUNCATED]",
    "bd": Object {
      "bca": "12345[...TRUNCATED]",
      "bcb": "12345[...TRUNCATED]",
    },
  },
}
`;
